### PR TITLE
Job Done!

### PR DESCRIPTION
**Pull request by IoghiStudio**

### CODE REVIEW

File: src/transformStateWithClones.js [hunk 1]
```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const currentState = { ...state };
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(currentState, obj.extraData);
```
Comment (mgrinko) [inline src/transformStateWithClones.js:15]: you mutate one initial cloned object all the time
It is better to create a copy on each iteration

**Review by mgrinko — APPROVED** [review]
Please finish your landing page BEFORE proceeding with Javascript Advanced!!!

**Review by mgrinko — CHANGES_REQUESTED** [review]
Please fix my previous comment

File: src/transformStateWithClones.js (hunk 1 above)
Comment (IoghiStudio) [reply src/transformStateWithClones.js:15]: Thanks for advice!

Comment (IoghiStudio) [reply src/transformStateWithClones.js:15]: I don't understand.

Comment (mgrinko) [reply src/transformStateWithClones.js:15]: You created a copy before the loop
and modify it on every action
But it is better to create a copy for each action 
I'll show you tomorrow

File: src/transformStateWithClones.js [hunk 2]
```diff
@@ -7,7 +7,34 @@
...
+        }
+      }
+    }
+
+    if (obj.type === 'clear') {
+      if (Object.keys(currentState).length !== 0) {
+        for (const key in currentState) {
+          delete currentState[key];
+        }
+      }
```
Comment (mgrinko) [inline src/transformStateWithClones.js:31]: If you need to get an empty object
Why do you copy the props?

**Review by mgrinko — CHANGES_REQUESTED** [review]

File: src/transformStateWithClones.js [hunk 3]
```diff
@@ -7,7 +7,34 @@
...
+  const currentState = { ...state };
+
+  for (const obj of actions) {
+    if (obj.type === 'addProperties') {
+      Object.assign(currentState, obj.extraData);
+    }
+
+    if (obj.type === 'removeProperties') {
+      for (const key of obj.keysToRemove) {
+        if (currentState.hasOwnProperty(key)) {
```
[the remark this answers is not in the document]
Comment (IoghiStudio) [reply src/transformStateWithClones.js:20]: That's right!

File: src/transformStateWithClones.js [hunk 4]
```diff
@@ -7,7 +7,28 @@
...
+      Object.assign(currentState, action.extraData);
+    }
+
+    if (type === 'removeProperties') {
+      action.keysToRemove.forEach(key => delete currentState[key]);
+    }
+
+    if (type === 'clear') {
+      currentState = {};
+    }
```
Comment (meforoma) [inline src/transformStateWithClones.js:26]: better use a `switch..case` construct that provides you with a `default` case, where you can provide an explanatory fallback, like
```
throw new Error(`${action} is not supported`)
```

**Review by meforoma — CHANGES_REQUESTED** [review]
see the comment about `switch..case`

**Pull request closed without merging**